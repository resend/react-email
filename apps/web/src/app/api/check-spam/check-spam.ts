import { parsePointingTableRows } from '@/utils/spam-assassin/parse-pointing-table-rows';
import { sendToSpamd } from '@/utils/spam-assassin/send-to-spamd';

export async function checkSpam(html: string, plainText: string) {
  const response = await sendToSpamd(html, plainText);

  const tableRows = parsePointingTableRows(response);

  const filteredRows = tableRows.filter(
    (row) =>
      !row.description.toLowerCase().includes('header') &&
      !row.ruleName.includes('HEADER') &&
      row.pts !== 0,
  );

  const checks = filteredRows.map((row) => ({
    name: row.ruleName,
    description: row.description,
    points: row.pts,
  }));

  const points = checks.reduce((acc, check) => acc + check.points, 0);

  return {
    checks,
    isSpam: points >= 5.0,
    points,
  };
}
