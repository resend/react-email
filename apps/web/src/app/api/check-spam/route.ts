import { type NextRequest, NextResponse } from 'next/server';
import { ZodError, z } from 'zod';
import { checkSpam } from './check-spam';

export const dynamic = 'force-dynamic';

export function OPTIONS() {
  return Promise.resolve(NextResponse.json({}));
}

const bodySchema = z.object({
  html: z.string(),
  plainText: z.string(),
});

export async function POST(req: NextRequest) {
  try {
    const { html, plainText } = bodySchema.parse(await req.json());

    return NextResponse.json(await checkSpam(html, plainText));
  } catch (exception) {
    if (exception instanceof Error) {
      return NextResponse.json(
        { error: exception.message },
        { status: exception instanceof ZodError ? 400 : 500 },
      );
    }

    return NextResponse.json(
      { error: 'Something went wrong' },
      { status: 500 },
    );
  }
}
