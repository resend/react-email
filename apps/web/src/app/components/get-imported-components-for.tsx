import { promises as fs } from 'node:fs';
import path from 'node:path';
import { parse } from '@babel/parser';
import traverse from '@babel/traverse';
import { render } from '@react-email/components';
import { z } from 'zod';
import { Layout } from '../../../components/_components/layout';
import type { Category, Component } from '../../../components/structure';
import {
  getComponentPathFromSlug,
  pathToComponents,
} from '../../../components/structure';

/**
 * Tailwind and Inline Styles are both with React, but the React
 * option is meant for where Tailwind nor Inline Styles are used
 * at all in the markup.
 */
export type CodeVariant = 'tailwind' | 'inline-styles' | 'react' | 'html';

export interface ImportedComponent extends Component {
  code: Partial<Record<CodeVariant, string>> & { html: string };
}

const ComponentModule = z.object({
  component: z.record(z.string(), z.any()),
});

const getComponentCodeFrom = (fileContent: string): string => {
  const parsedContents = parse(fileContent, {
    sourceType: 'unambiguous',
    strictMode: false,
    errorRecovery: true,
    plugins: ['jsx', 'typescript', 'decorators'],
  });

  let componentCode: string | undefined;
  traverse(parsedContents, {
    VariableDeclarator({ node }) {
      if (
        node.id.type === 'Identifier' &&
        node.id.name === 'component' &&
        (node.init?.type === 'JSXElement' || node.init?.type === 'JSXFragment')
      ) {
        const expression = node.init;
        if (expression.start !== null && expression.end !== null) {
          componentCode = fileContent.slice(expression.start, expression.end);
        }
      }
    },
  });

  if (!componentCode) {
    throw new Error('Could not find the source code for the component');
  }

  return componentCode
    .split(/\r\n|\r|\n/)
    .map((line) => line.replace(/^\s{2}/, ''))
    .join('\n');
};

export const getComponentElement = async (
  filepath: string,
): Promise<React.ReactElement> => {
  const relativeFilepath = path.relative(pathToComponents, filepath);
  const patternModule = ComponentModule.parse(
    await import(
      `../../../components/${relativeFilepath.replace(
        path.extname(relativeFilepath),
        '',
      )}`
    ),
  );
  return patternModule.component as React.ReactElement;
};

export const getImportedComponent = async (
  component: Component,
): Promise<ImportedComponent> => {
  const dirpath = getComponentPathFromSlug(component.slug);
  const variantFilenames = await fs.readdir(dirpath);

  if (variantFilenames.length === 1 && variantFilenames[0] === 'index.tsx') {
    const filePath = path.join(dirpath, 'index.tsx');
    const element = <Layout>{await getComponentElement(filePath)}</Layout>;
    const html = await render(element, {
      pretty: true,
    });
    const fileContent = await fs.readFile(filePath, 'utf8');
    const code = getComponentCodeFrom(fileContent);
    return {
      ...component,
      code: {
        react: code,
        html,
      },
    };
  }

  const codePerVariant: ImportedComponent['code'] = { html: '' };

  const elements = await Promise.all(
    variantFilenames.map(async (variantFilename) => {
      const filePath = path.join(dirpath, variantFilename);
      return getComponentElement(filePath);
    }),
  );

  const fileContents = await Promise.all(
    variantFilenames.map(async (variantFilename) => {
      const filePath = path.join(dirpath, variantFilename);
      return fs.readFile(filePath, 'utf8');
    }),
  );

  variantFilenames.forEach((variantFilename, index) => {
    const variantKey = variantFilename.replace('.tsx', '') as CodeVariant;
    codePerVariant[variantKey] = getComponentCodeFrom(fileContents[index]);
  });

  const element = <Layout>{elements[0]}</Layout>;

  codePerVariant.html = await render(element, {
    pretty: true,
  });

  return {
    ...component,
    code: codePerVariant,
  };
};

export const getImportedComponentsFor = async (
  category: Category,
): Promise<ImportedComponent[]> => {
  return Promise.all(
    category.components.map((component) => getImportedComponent(component)),
  );
};
