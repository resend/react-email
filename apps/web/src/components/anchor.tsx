import classNames from 'classnames';
import type * as React from 'react';

export const Anchor: React.FC<
  Readonly<React.ComponentPropsWithoutRef<'a'>>
> = ({ className, ...props }) => (
  <a
    className={classNames(
      'rounded-sm outline-none transition-transform duration-200 ease-in-out',
      'hover:-translate-y-1',
      'focus:ring-2 focus:ring-white/20 focus:ring-offset-4 focus:ring-offset-black',
      'text-slate-12',
      className,
    )}
    {...props}
  >
    {props.children}
  </a>
);
