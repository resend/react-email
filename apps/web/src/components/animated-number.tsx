'use client';

import { motion, useInView, useSpring, useTransform } from 'framer-motion';
import { useEffect, useRef } from 'react';

interface AnimatedNumberProps {
  from?: number;
  to?: number;
  delay?: number;
  className?: string;
}

export const AnimatedNumber = ({
  from = 0,
  to = 100,
  delay = 1,
  className = '',
}: AnimatedNumberProps) => {
  const ref = useRef(null);
  const isInView = useInView(ref, { once: true });

  const spring = useSpring(from, { mass: 0.8, stiffness: 75, damping: 15 });
  const display = useTransform(spring, (current) =>
    Math.round(current).toLocaleString(),
  );

  useEffect(() => {
    if (isInView) {
      const timer = setTimeout(() => {
        spring.set(to);
      }, delay * 1000);

      return () => clearTimeout(timer);
    }
  }, [isInView, spring, to, delay]);

  return (
    <motion.span ref={ref} className={className}>
      {display}
    </motion.span>
  );
};
