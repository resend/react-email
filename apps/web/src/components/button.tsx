import { Slot } from '@radix-ui/react-slot';
import classNames from 'classnames';
import * as React from 'react';
import { unreachable } from '../utils/unreachable';

type Appearance = 'white' | 'gradient';
type Size = '1' | '2' | '3' | '4';

interface ButtonProps extends React.ComponentPropsWithoutRef<'button'> {
  asChild?: boolean;
  appearance?: Appearance;
  size?: Size;
}

export const Button = React.forwardRef<
  HTMLButtonElement,
  Readonly<ButtonProps>
>(
  (
    {
      asChild,
      appearance = 'white',
      className,
      children,
      size = '2',
      ...props
    },
    forwardedRef,
  ) => {
    const Comp = asChild ? Slot : 'button';

    return (
      <Comp
        className={classNames(
          getSize(size),
          getAppearance(appearance),
          'inline-flex items-center justify-center border font-medium',
          className,
        )}
        ref={forwardedRef}
        type="button"
        {...props}
      >
        {children}
      </Comp>
    );
  },
);

Button.displayName = 'Button';

const getAppearance = (appearance: Appearance) => {
  switch (appearance) {
    case 'white':
      return [
        'bg-white text-black',
        'hover:bg-white/90',
        'focus:ring-slate-8 focus:bg-white/90 focus:outline-none focus:ring-2',
      ];
    case 'gradient':
      return [
        'bg-gradient border-[#34343A] backdrop-blur-[1.25rem]',
        'hover:bg-gradientHover',
        'focus:bg-gradientHover focus:outline-none focus:ring-2 focus:ring-white/20',
      ];
    default:
      unreachable(appearance);
  }
};

const getSize = (size: Size) => {
  switch (size) {
    case '1':
      return '';
    case '2':
      return 'h-8 gap-2 rounded-md px-3 text-[.875rem] transition-colors';
    case '3':
      return 'h-10 gap-2 rounded-md px-4 text-[.875rem] transition-colors';
    case '4':
      return 'h-11 gap-2 rounded-md px-4 text-base transition-colors';
    default:
      unreachable(size);
  }
};
