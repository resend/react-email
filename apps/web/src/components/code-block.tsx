import classNames from 'classnames';
import type { Language } from 'prism-react-renderer';
import { Highlight } from 'prism-react-renderer';
import * as React from 'react';

const theme = {
  plain: {
    color: '#EDEDEF',
    fontSize: 13,
    fontFamily: 'CommitMono, monospace',
  },
  styles: [
    {
      types: ['comment'],
      style: {
        color: '#706F78',
      },
    },
    {
      types: ['atrule', 'keyword', 'attr-name', 'selector'],
      style: {
        color: '#7E7D86',
      },
    },
    {
      types: ['punctuation', 'operator'],
      style: {
        color: '#706F78',
      },
    },
    {
      types: ['class-name', 'function', 'tag', 'key-white'],
      style: {
        color: '#EDEDEF',
      },
    },
  ],
};

interface CodeBlockProps {
  children: string;
  className?: string;
  language?: Language;
}

export const CodeBlock: React.FC<Readonly<CodeBlockProps>> = ({
  children,
  language = 'html',
}) => {
  const value = children.trim();

  return (
    <Highlight code={value} language={language} theme={theme}>
      {({ tokens, getLineProps, getTokenProps }) => (
        <>
          <div
            className="absolute right-0 top-0 h-px w-[12.5rem]"
            style={{
              background:
                'linear-gradient(90deg, rgba(56, 189, 248, 0) 0%, rgba(56, 189, 248, 0) 0%, rgba(232, 232, 232, 0.2) 33.02%, rgba(143, 143, 143, 0.6719) 64.41%, rgba(236, 72, 153, 0) 98.93%)',
            }}
          />

          <pre className="p-4 font-mono">
            {tokens.map((line, i) => {
              const lineProps = getLineProps({ line, key: i });

              return (
                <div
                  key={i}
                  {...lineProps}
                  className={classNames('whitespace-pre', {
                    "before:mr-2 before:text-slate-11 before:content-['$']":
                      language === 'bash' && tokens.length === 1,
                  })}
                >
                  {line.map((token, key) => {
                    const tokenProps = getTokenProps({ token, key });

                    const isException =
                      token.content === 'from' &&
                      line[key + 1]?.content === ':';
                    token.types = isException
                      ? [...token.types, 'key-white']
                      : token.types;

                    return (
                      <React.Fragment key={key}>
                        <span {...tokenProps} />
                      </React.Fragment>
                    );
                  })}
                </div>
              );
            })}
          </pre>
          <div
            className="absolute bottom-0 left-0 h-px w-[12.5rem]"
            style={{
              background:
                'linear-gradient(90deg, rgba(56, 189, 248, 0) 0%, rgba(56, 189, 248, 0) 0%, rgba(232, 232, 232, 0.2) 33.02%, rgba(143, 143, 143, 0.6719) 64.41%, rgba(236, 72, 153, 0) 98.93%)',
            }}
          />
        </>
      )}
    </Highlight>
  );
};
