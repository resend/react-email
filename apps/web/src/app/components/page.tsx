import classNames from 'classnames';
import type { Metadata } from 'next';
import dynamic from 'next/dynamic';
import Link from 'next/link';
import React from 'react';
import { componentsStructure } from '../../../components/structure';
import PageTransition from '../../components/page-transition';
import { Spotlight } from '../../components/spotlight';
import { slugify } from '../../utils/slugify';

const title = 'Components - React Email';
const description =
  'Build beautiful emails with pre-built components that you can copy-and-paste into your app.';

export const metadata: Metadata = {
  title,
  description,
  openGraph: {
    title,
    description,
    images: [
      {
        url: 'https://react.email/static/covers/patterns.png',
      },
    ],
  },
  alternates: {
    canonical: '/components',
  },
};

const ComponentsPage = async () => (
  <>
    <div className="pointer-events-none absolute inset-0 flex justify-center">
      <div className="hidden h-full w-full max-w-7xl grid-cols-2 gap-4 px-4 lg:grid">
        <div className="border-l border-l-slate-4 border-r-slate-3" />
        <div className="border-r border-r-slate-4" />
      </div>
    </div>
    <PageTransition className="pb-10" key="about" tag="main">
      <div className="flex w-full flex-col gap-2 px-6 pb-10 pt-16 md:px-8">
        <h1 className="text-2xl font-bold text-slate-12">Components</h1>
        <p>
          Build beautiful emails with pre-built components that you can
          copy-and-paste into your app.
        </p>
      </div>
      <div className="relative grid grid-cols-1 gap-x-4 px-1 pb-10 md:grid-cols-2 md:px-0 lg:grid-cols-3">
        <div className="absolute left-1/2 top-0 h-px w-[100dvw] -translate-x-1/2 border-t border-slate-4" />
        <div className="absolute bottom-0 left-1/2 h-px w-[100dvw] -translate-x-1/2 border-b border-slate-4" />
        {componentsStructure.map((category, index) => {
          const slug = slugify(category.name);
          const Illustration = dynamic(
            () =>
              import(
                `@/illustrations/${category.name
                  .toLowerCase()
                  .replace(/ /g, '-')}`
              ),
          );

          return (
            <Link
              className={classNames(
                'group relative isolate mt-7 cursor-pointer scroll-m-6 rounded-md focus:outline-none focus:ring focus:ring-slate-2 md:before:absolute md:before:inset-0 md:before:rounded-md md:before:border md:before:border-dashed md:before:border-slate-4 md:before:transition-colors md:before:duration-[720ms] md:before:ease-[cubic-bezier(.24,.9,.32,1.4)] md:hover:before:border-slate-6 md:focus:before:border-slate-6',
                {
                  'md:ml-6': index % 3 === 0,
                  'md:mx-3': index % 3 === 1,
                  'md:mr-6': index % 3 === 2,
                },
              )}
              href={`/components/${slug}`}
              key={category.name}
              tabIndex={0}
            >
              <Spotlight
                className={classNames(
                  'relative isolate flex cursor-pointer flex-col justify-end rounded-md bg-black p-4 group-focus:ring group-focus:ring-slate-2 md:transition-transform md:duration-[240ms] md:ease-[cubic-bezier(.36,.66,.6,1)]',
                  {
                    'md:group-hover:-translate-x-2 md:group-hover:-translate-y-2 md:group-focus:-translate-x-2 md:group-focus:-translate-y-2':
                      index % 3 === 0,
                    'md:group-hover:-translate-y-2 md:group-focus:-translate-y-2':
                      index % 3 === 1,
                    'md:group-hover:-translate-y-2 md:group-hover:translate-x-2 md:group-focus:-translate-y-2 md:group-focus:translate-x-2':
                      index % 3 === 2,
                  },
                )}
              >
                <div className="pointer-events-none absolute inset-0 rounded-md border border-slate-4 transition-colors duration-300 ease-[cubic-bezier(.36,.66,.6,1)] group-hover:border-slate-6 group-focus:border-slate-6" />
                <div className="relative flex aspect-[2/1] items-center justify-center overflow-hidden rounded-sm text-slate-300">
                  <div className="absolute inset-0 bg-transparent bg-[radial-gradient(#27272A_.0313rem,transparent_.0313rem),_radial-gradient(#27272A_.0313rem,transparent_.0313rem)] opacity-80 [background-position:0_0,.625rem_.625rem] [background-size:1.25rem_1.25rem]" />
                  <Illustration />
                </div>
                <h3 className="relative z-[2] mt-4 font-semibold capitalize leading-7 tracking-wide text-slate-12">
                  {category.name}
                </h3>
                <span className="relative z-[2] text-xs text-slate-11">
                  {category.components.length} component
                  {category.components.length > 1 && 's'}
                </span>
              </Spotlight>
            </Link>
          );
        })}
      </div>
    </PageTransition>
  </>
);

export default ComponentsPage;
