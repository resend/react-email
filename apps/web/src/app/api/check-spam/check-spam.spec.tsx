import { render } from '@react-email/components';
import { checkSpam } from './check-spam';

test('checkSpam()', async () => {
  const template = (
    <html lang="en">
      <body>
        This email is spam. We sell rolexss for cheap. Get your viagra. We also
        sell weight loss pills today. Money back guaranteed. sEnd me $500 and
        I'll send you back $700. don't tell anyone. Send this email to ten
        friends
      </body>
    </html>
  );
  const html = await render(template);
  // const plainText = await render(template, { plainText: true });
  const plainText = 'Completely different content from the original';

  expect(await checkSpam(html, plainText)).toMatchSnapshot();
});
