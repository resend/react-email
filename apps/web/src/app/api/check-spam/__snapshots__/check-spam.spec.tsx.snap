// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`checkSpam() > with most spammy email 1`] = `
{
  "checks": [
    {
      "description": "BODY: Money back guarantee",
      "name": "MONEY_BACK",
      "points": 2.5,
    },
    {
      "description": "BODY: HTML and text parts are different",
      "name": "MPART_ALT_DIFF",
      "points": 0.7,
    },
    {
      "description": "Refers to an erectile drug",
      "name": "DRUGS_ERECTILE",
      "points": 2.2,
    },
  ],
  "isSpam": true,
  "points": 5.4,
}
`;

exports[`checkSpam() > with real email template 1`] = `
{
  "checks": [],
  "isSpam": false,
  "points": 0,
}
`;
